### PR TITLE
#167969811 Users can subscribe to articles

### PR DESCRIPTION
**Pull request by ibkadeeko**
#### What does this PR do?
This pull request helps users subscribe to articles

#### Description of Task to be completed
- user can create article subscriptions

#### How should this be manually tested?
- clone this branch
- run `npm install`
- using postman send a `POST` request to `/api/v1/users/subscriptions`

#### Any Background Context You want to provide?
N/A

#### What are the relevant pivotal tracker stories?

[#167969811](https://pivotaltracker.com/story/show/167969811)

#### Screenshots (if appropriate)
Screenshot 2019-08-23 at 15 33 35
Screenshot 2019-08-23 at 21 01 52

#### Questions:
N/A

### CODE REVIEW

**Review by Jonathan4github — APPROVED** [review]
Good job

**Comment by iverenshaguy** [comment]
Can we have the category name quoted in the error message i.e `Category "safety" not supported`? The current message might be confusing. Also, I think it would be nice for the category objects in the response to also include the category name/title. That way, we are not wondering what id belongs to what category. Thanks

File: src/controllers/user.js [hunk 1]
```diff
@@ -1,31 +1,28 @@
...
   /**
-* Generic update profile
-* @function
-*
-* @param {Object} request - express request object
-* @param {Object} response - express response object
-*
-* @returns {Object} - callback that execute the controller
-*/
+   * Generic update profile
```
Comment (iverenshaguy) [inline src/controllers/user.js]: I know I told you guys not to use descriptions like this, in the codebase. It's not really nice still seeing stuff like this. I know it's not your code but could you please help update it? Thank you

File: src/controllers/user.js [hunk 2]
```diff
@@ -60,18 +59,20 @@ export default {
...
     delete existingUser.password;
 
-    return response.status(200).json({ status: 'success', message: 'Request was successful', data: existingUser });
+    return response
+      .status(200)
+      .json({ status: 'success', message: 'Request was successful', data: existingUser });
   },
 
   /**
-    * A user can follow and unfollow a user
```
Comment (iverenshaguy) [inline src/controllers/user.js:67]: Thanks for fixing all these alignment issues

File: src/controllers/user.js [hunk 3]
```diff
@@ -60,18 +59,20 @@ export default {
...
-    * @function
-    *
-    * @param {Object} request - express request object
-    * @param {Object} response - express response object
-    *
-    * @returns {Object} - callback that execute the controller
-    */
+   * A user can follow and unfollow a user
```
Comment (iverenshaguy) [inline src/controllers/user.js]: Please help update to something more descriptive

File: src/controllers/user.js [hunk 4]
```diff
@@ -171,7 +172,38 @@ export default {
       data,
       count,
       page,
-      message: 'Request successful',
+      message: 'Request successful'
+    });
+  },
+
+  /**
+   * Controller to create subscriptions for ser
```
Comment (iverenshaguy) [inline src/controllers/user.js]: Could you please complete this description? Thanks

File: src/models/Users.js [hunk 5]
```diff
@@ -317,4 +317,30 @@ export default class Users extends Model {
...
+   * @static
+   * @memberof Users
+   *
+   * @param {Array} categoryIds Array of category Ids to be subscribed to
+   *
+   * @returns {Object} Subscription data
+   */
+  async bulkUpsertSubscriptions(categoryIds) {
+    let data;
+    data = categoryIds.map(async (categoryId) => {
```
Comment (iverenshaguy) [inline src/models/Users.js]: Fun fact, you can also use reduce for this. You knowhow reduce has to wait for the previous execution to be completed before continuing, you can leverage that behaviour. Not saying you should change this though, it's just another option

File: src/models/Users.js [hunk 6]
```diff
@@ -317,4 +317,30 @@ export default class Users extends Model {
...
+      const isExists = await this.getSubscriptions({ where: { categoryId } }, { raw: true });
+      if (isExists.length) {
+        return isExists[0];
+      }
+      const { dataValues } = await this.createSubscription({ categoryId });
```
Comment (iverenshaguy) [inline src/models/Users.js]:
```suggestion

      const { dataValues } = await this.createSubscription({ categoryId });
```

File: tests/routes/profile.test.js [hunk 7]
```diff
@@ -250,4 +268,93 @@ describe('Profile', () => {
...
+    it('should get throw error if no categories are supplied', async () => {
+      ({ token } = responseToken.body.data);
+      const response = await chai
+        .request(app)
+        .post(url)
+        .set('authorization', `Bearer ${token}`);
+
+      response.should.have.status(400);
+      response.body.status.should.eql('error');
+      response.body.error.errors.categories.should.eql('Categories cannot be empty');
```
Comment (iverenshaguy) [inline tests/routes/profile.test.js:285]: Thanks

File: tests/routes/profile.test.js [hunk 8]
```diff
@@ -250,4 +268,93 @@ describe('Profile', () => {
...
+    it('should successfully subscribe to articles', async () => {
+      const response = await chai
+        .request(app)
+        .post(url)
+        .set('authorization', `Bearer ${token}`)
+        .send({ categories: ['motivation', 'time management'] });
+
+      response.should.have.status(200);
+      response.body.status.should.eql('success');
+      response.body.message.should.eql('Successfully subscribed to motivation, time management');
```
Comment (iverenshaguy) [inline tests/routes/profile.test.js:357]: Thanks for your descriptive tests

**Review by iverenshaguy — APPROVED** [review]
Really good job, please review my comments so that we can merge

**Comment by ibkadeeko** [comment]
> Can we have the category name quoted in the error message i.e `Category "safety" not supported`? The current message might be confusing. Also, I think it would be nice for the category objects in the response to also include the category name/title. That way, we are not wondering what id belongs to what category. Thanks

I have changed the error response. The success response is what is returned from the database and the category name is not in the subscription database table

**Comment by iverenshaguy** [comment]
> > Can we have the category name quoted in the error message i.e `Category "safety" not supported`? The current message might be confusing. Also, I think it would be nice for the category objects in the response to also include the category name/title. That way, we are not wondering what id belongs to what category. Thanks
> 
> I have changed the error response. The success response is what is returned from the database and the category name is not in the subscription database table

Please eager load the category name. Thanks

**Review by successgilli — APPROVED** [review]

**Review by iverenshaguy — APPROVED** [review]
Really good job

**Pull request merged**